[Liu Yongliang] iP

tlylt opened the pull request.

Commit:
Add Gradle support

Commit:
build.gradle: Update version to 8.29

tlylt committed:
Greet, Echo, Exit

tlylt committed:
Add, List

tlylt committed:
Mark as Done

tlylt committed:
ToDo, Event, Deadline

tlylt committed:
Automated Text UI Testing

tlylt committed:
Level-5: handle errors

Input from users may not follow recognizable pattern.

Add exceptions from HahaException class to handle undesirable input.

Include exception classes for better organization of errors.

This also completes A-Exceptions extension.

tlylt committed:
Update test

tlylt committed:
Level-6 Delete

Add delete function to remove specified task.

Fix test typos.

Extensions: A-Collections, A-Exceptions and A-Inheritance are completed.

tlylt committed:
Major Revamp: A-Enums and Code refactor

All related code to storing of tasks and parsing of input are within the
class Haha

Restructure and create the following classes:
- TaskType enum to specify all possible task types
- LegitCommand enum to specify all accepted commands
- Parser class to deal with sanitizing input
- Database class to deal with modifying of task records
- A few specific exceptions that inherit from HahaException

Extra:
- Still not happy with code structure within subclass
- May decide to modify code in Parser for better organization
and fit OOP style

tlylt committed:
Fix switch case code style

tlylt committed:
Add save to disk functionality

Data about tasks are stored in memory and will not preserve to the next
section.

Let's write existing data about user's todo into a file under
current working directory and into /data/database.txt

This is to ensure on starting up the program again, data can be read
from the file and display to users.

tlylt committed:
Parse date and time for user input

Date and time are treated as pure string, which cannot be parsed and
compared later on.

Let's parse the input string and store date and time information as
LocalDate and LocalTime object. Also formatting is applied when printing
the information.

This helps users to see the date and time information easily.

tlylt committed:
Merge branch 'branch-Level-7' into master

tlylt committed:
Merge branch 'branch-Level-8' into master

tlylt committed:
Add more OOP: Ui, storage

tlylt committed:
Make the code more OOP

Different components are within the Haha class and  this might cause
confusion and increase difficulty in managing code.

Let's apply OOP style and extract out four classes: Ui, Storage, Parser
and TaskList.

This is to separate different logic into different files.

tlylt committed:
Divide classes into packages

Classes not organized.

Let's package all files into a single package, tlylt.haha.

This helps for code organisation.

tlylt committed:
Add Junit tests

Current program only tested with text-ui-test, which may not be enough.

Let's add Junit tests to increase testing and hence ensure program is
correct.

Test cases for Ui and Parser classes are added to ensure they work as
expected.

tlylt committed:
Package the App as a JAR file

Program needs to be compiled to run and this is not user friendly.

Let's fix the storage and package up the program for release as a JAR
file.

This allows users to download and use the program easily.

tlylt committed:
Find related tasks

tlylt committed:
Follow coding standard

tlylt committed:
Add JavaDoc comments

tlylt committed:
Merge branch 'branch-A-JavaDoc' into master

tlylt committed:
Merge branch 'branch-A-CodingStandard' into master

# Conflicts:
#	src/main/java/tlylt/haha/LegitCommand.java
#	src/main/java/tlylt/haha/Storage.java
#	src/main/java/tlylt/haha/Task.java

tlylt committed:
Merge branch 'branch-Level-9' into master

# Conflicts:
#	src/main/java/tlylt/haha/TaskList.java
#	src/main/java/tlylt/haha/Ui.java

tlylt committed:
Fix minor merge conflicts

wangtao0717 commented on src/main/java/tlylt/haha/TaskList.java:
```diff
@@ -104,13 +104,6 @@ void listFromDB() {
-=======
     private void findFromDB(String keyword, Ui ui) {
         System.out.println("Here are the matching tasks in your list:");
         boolean hasRelated = false;
```
Perhaps use variable name isRelated is more  sounded  boolean.

tlylt replied:
Hi Wang Tao, thank you for your review.
In this case, I am referring to whether "database" **has related** tasks instead of whether a task **is related** to the search string, hence I think "hasRelated" is perhaps more grammatically correct in this sense?
Nonetheless, I would like to thank you for your time and I will check if my other boolean values could be better named.

wangtao0717 commented on src/main/java/tlylt/haha/LegitCommand.java:
```diff
@@ -10,7 +10,9 @@
-    BYE("bye");
+    BYE("bye"),
+    FIND("find");
+    
```
I like your usage of enum

wangtao0717 reviewed (commented):
Overall, I found your code is concise and  easy to read. The structure is clear and it basically follows the code standard.
So great!

tlylt committed:
Merge branch 'add-gradle-support' into A-Gradle

tlylt committed:
Automate project builds using Gradle

Without Gradle support, the build needs to be built manually.

Let's add Gradle support to automate build processes.

This allows for automation in building and testing.

tlylt committed:
Merge branch 'A-Gradle' into master

tlylt committed:
Use CheckStyle

Manual checking of code style violations can be ineffective.

Let's use Gradle to add checkstyle support.

This allows running of checkstyle as a Gradle task.

tlylt committed:
Merge branch 'A-CheckStyle' into master

tlylt committed:
Add a GUI to Haha

Without GUI, the display can be quite limited.

Let's add a GUI using JavaFX.

Users can now type into the input box and see a response in a chat bot
style.

tlylt committed:
Merge branch 'branch-Level-10' into master

sylviaokt commented on src/main/java/tlylt/haha/DialogBox.java:
```diff
@@ -0,0 +1,55 @@
+
+public class DialogBox extends HBox {
+    private Label text;
+    private ImageView displayPicture;
```
Perhaps this variable could be named using a noun instead?

sylviaokt commented on src/main/java/tlylt/haha/Parser.java:
```diff
@@ -0,0 +1,209 @@
+     * @return Correct task number as an integer.
+     * @throws HahaTaskNumberNotIntException Task number given is not an integer.
+     */
+    static int taskNumber(String command) throws HahaTaskNumberNotIntException {
```
Perhaps this method could be named getTaskNumber, which is a verb, to better represent the method?

sylviaokt commented on src/main/java/tlylt/haha/Parser.java:
```diff
@@ -0,0 +1,209 @@
+        String[] tokens = line.split("\\|");
+        String type = tokens[0];
+        boolean isDone = tokens[1].equals("true");
+        String description = tokens[2];
```
Perhaps you could be more specific as to which description is it for?

sylviaokt commented on src/main/java/tlylt/haha/Storage.java:
```diff
@@ -0,0 +1,47 @@
+ * Utility for storing tasks to file.
+ */
+public class Storage {
+    private Path file;
```
Perhaps specify filePath instead, because file might sound misleading?

sylviaokt commented on src/main/java/tlylt/haha/Storage.java:
```diff
@@ -0,0 +1,47 @@
+ * Utility for storing tasks to file.
+ */
+public class Storage {
+    private Path file;
+
+    /**
+     * Initialises by checking and creating storage file if needed.
+     */
+    public Storage() {
+        try {
+            String currentDir = System.getProperty("user.dir");
+            Path folder = Paths.get(currentDir, "Haha_data");
```
Perhaps specify filePath and folderPath instead? Otherwise it might sound misleading.

sylviaokt commented on src/main/java/tlylt/haha/Task.java:
```diff
@@ -0,0 +1,106 @@
+ * Abstract representation of a task.
+ */
+public abstract class Task {
+    protected final String description;
+    protected boolean isDone;
+    protected final String type;
```
Perhaps for the variables description and type, you could be more specific?

tlylt replied:
ERM, Perhaps the comments could have been done in one single review, following the guide? Now my mailbox is bombed 💣

sylviaokt commented on src/main/java/tlylt/haha/TaskList.java:
```diff
@@ -0,0 +1,220 @@
+     *
+     * @param task Task created by user.
+     */
+    String addToDB(Task task) {
+        String response = "";
+
+        database.add(task);
+        response += tellAdd();
+        System.out.println("  " + database.get(database.size() - 1));
+        response += "  " + database.get(database.size() - 1) + "\n";
+        response += tellSize();
+        return response;
+    }
+
+    /**
+     * Removes task from database.
+     *
+     * @param inputNum String that will be parsed for task number.
+     */
+    String deleteFromDB(String inputNum) {
```
Perhaps spell out database? DB could be confusing.

sylviaokt commented on src/main/java/tlylt/haha/TaskList.java:
```diff
@@ -0,0 +1,220 @@
+    /**
+     * Updates database to file.
+     */
+    void updateFile() {
+        List<String> str = new ArrayList<>();
+        database.forEach(task -> str.add(task.fileStorageFormat()));
+        try {
+            Files.write(Paths.get(System.getProperty("user.dir"), "Haha_data", "database.txt"),
+                    str, StandardOpenOption.TRUNCATE_EXISTING);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
```
I like how your method names in general uses verb, and accurately represents their functionality.

sylviaokt commented:
Overall, this code followed the Java coding quality guidelines well, keep it up! The code was easy to read, and the naming issues I have pointed out are only minor issues, as the rest of the naming are accurate and good!

tlylt committed:
Use Assertions

Without assertions, assumptions cannot be checked in switch statements.

Let's add in assert statements in default branch of switch statements.

If the code runs through the default branch, the assumption is wrong and
hence assertion error is invoked.

tlylt committed:
Improve code quality

Current code requires refactoring due to the newly added GUI.

Let's improve code quality by refactoring some of the methods.

This makes the maintenance of code easier.

tlylt committed:
Improve UI style

tlylt committed:
Use Lambdas

Named functions can be harder to follow for a 1-liner function.

Let's use Lambdas in configuring  button interactions.

Now it is easier to see that on click button will trigger the handling
function for inputs.

tlylt committed:
Use Streams.

For loops can be harder to trace as compared to streams.

Let's use streams in identifying keyword during parsing.

This is easier to see what are the steps of parsing.

tlylt committed:
Merge pull request #1 from tlylt/branch-A-Assertions

Use Assertions

tlylt committed:
Merge branch 'master' into branch-A-CodeQuality

tlylt committed:
Merge pull request #2 from tlylt/branch-A-CodeQuality

Improve code quality

tlylt committed:
Merge branch 'master' into branch-A-Lambdas

tlylt committed:
Merge branch 'master' into branch-A-Streams

tlylt committed:
Merge pull request #3 from tlylt/branch-A-Lambdas

Use Lambdas

tlylt committed:
Merge branch 'master' into branch-A-Streams

tlylt committed:
Merge pull request #4 from tlylt/branch-A-Streams

Use Streams

tlylt committed:
Set up CI

Program is not continously integrated.

Let's use Github Actions to set up Continuous Integeration.

Gradle check will be triggered for updates.

tlylt committed:
Fix checkstyle violations

tlylt committed:
Fix trailing space

tlylt committed:
Remove outdated text IO test

tlylt committed:
Add the ability to deal with duplicate tasks

Adding the same task may not be useful.

Let's detect if a todo/deadline/event about to be added already exist.

Warn the user about the duplicate and prevent it from being added.

tlylt committed:
Merge branch 'branch-C-DetectDuplicates' into master

tlylt committed:
Fix checkstyle violations

tlylt committed:
Improve the GUI

GUI is not clean and nice enough.

Let's clean up the dialog box and also add in a background image.

This way the App looks more polished.

tlylt committed:
Clean up minor issues and code style violations

tlylt committed:
Set theme jekyll-theme-tactile

tlylt committed:
Fix edge cases

Wrong format of deadline and event tasks trigger undetected exceptions.

Let's create a new exception to handle it.

Correct response will be given to user to notify about format.

tlylt committed:
Merge branch 'master' of https://github.com/tlylt/ip into master

tlylt committed:
Update getTaskNumber test

tlylt committed:
Add a brief user guide

Commands can be confusing if users are using Haha for the first time.

Let's add a brief user guide to demonstrate all the capabilities.

Relevent commands and sample screenshots are included.

tlylt committed:
Fix minor typo

tlylt committed:
Clean up magic numbers

tlylt committed:
Fix gradle config
